agent/grpc/public/services/serverdiscovery/watch_servers.go

```diff
@@ -0,0 +1,157 @@
+package serverdiscovery
+
+import (
+	"context"
+	"errors"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/autopilotevents"
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/grpc/public"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/proto-public/pbserverdiscovery"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// We tag logs with a unique identifier to ease debugging. In the future this
+// should probably be an Open Telemetry trace ID.
+func streamID() string {
```

Comment:
We can re-use this helper I added under utils: https://github.com/hashicorp/consul/blob/main/agent/grpc/public/utils.go#L7